File: java/src/main/java/com/github/algo/strings/IntegerToRoman.java
package com.github.algo.strings;

import java.util.*;

public class IntegerToRoman {
    static Map<Integer, Character> romanNumbers = new HashMap<>();
    static List<Integer> factors = Arrays.asList(1000, 500, 100, 50, 10, 5, 1);
    static Map<Integer, Integer> subFactors = new HashMap<>();

    static {
        romanNumbers.put(1, 'I');
        romanNumbers.put(5, 'V');
        romanNumbers.put(10, 'X');
        romanNumbers.put(50, 'L');
        romanNumbers.put(100, 'C');
        romanNumbers.put(500, 'D');
        romanNumbers.put(1000, 'M');
        subFactors.put(5, 1);
        subFactors.put(10, 1);
        subFactors.put(50, 10);
        subFactors.put(100, 10);
        subFactors.put(500, 100);
        subFactors.put(1000, 100);
    }

    boolean isSubFact(Integer num, Integer factor) {
        return subFactors.containsKey(factor) && num - (num % subFactors.get(factor)) + subFactors.get(factor) == factor;
    }

    public String intToRoman(int num) {
        StringBuilder roman = new StringBuilder();
        while (num > 0) {
            for (int i = 0; i < factors.size(); i++) {
                Integer factor = factors.get(i);
                if (num >= factor) {
                    int count = num / factor;
                    num = num % factor;
                    Character ch = romanNumbers.get(factor);
                    roman.append(String.valueOf(ch).repeat(Math.max(0, count)));
                    break;
                }
                if (this.isSubFact(num, factor)) {
                    roman.append(romanNumbers.get(subFactors.get(factor)));
                    roman.append(romanNumbers.get(factor));
                    num = num % subFactors.get(factor);
                    break;
                }
            }
//            System.out.println(num + ", " + roman.toString());
        }
        return roman.toString();
    }

}

File: python/main/math/IntegerToRoman.py
# https://leetcode.com/problems/integer-to-roman/

FACTORS = [
    # 100_000,
    # 50_000,
    # 10_000,
    # 5000,
    1000,
    500,
    100,
    50,
    10,
    5,
    1
]
ROMAN_NUMBERS = {
    # 100_000: "Q",
    # 50_000: "W",
    # 10000: "Y",
    # 5000: "Z",
    1000: "M",
    500: "D",
    100: "C",
    50: "L",
    10: "X",
    5: "V",
    1: "I"
}
SUB_FACTOR = {
    # 100_000: 10000,
    # 50_000: 10000,
    # 10000: 1000,
    # 5000: 1000,
    1000: 100,
    500: 100,
    100: 10,
    50: 10,
    10: 1,
    5: 1
}


class Solution:

    def intToRoman(self, num: int) -> str:
        output = []
        while num > 0:
            for factor in FACTORS:
                if num >= factor:
                    count = max(num // factor, 0)
                    num = num % factor
                    output.append(ROMAN_NUMBERS[factor] * count)
                    break

                if factor in SUB_FACTOR:
                    subFactor = SUB_FACTOR[factor]
                    if num - (num % subFactor) == factor - subFactor:
                        num = num % (factor - subFactor)
                        output.append(ROMAN_NUMBERS[subFactor] + ROMAN_NUMBERS[factor])
                        break
        # print(output)
        return "".join(output)

        pass


if __name__ == "__main__":
    s = Solution()
    for i in range(1, 100):
        roman = s.intToRoman(i)
        print(i, "-", roman)
